server/src/main/java/org/opensearch/indices/SystemIndexRegistry.java

```diff
@@ -56,14 +57,26 @@ static void register(Map<String, Collection<SystemIndexDescriptor>> pluginAndMod
             .collect(Collectors.toList());
         descriptors.add(TASK_INDEX_DESCRIPTOR);
 
+        SYSTEM_INDEX_DESCRIPTORS_MAP = descriptorsMap;
```

Comment:
@cwperks sorry, I don't understand why we need SYSTEM_INDEX_DESCRIPTORS and SYSTEM_INDEX_DESCRIPTORS_MAP ?

Reply:
Its an optimization so that the data structures in this class do not need to be repeatedly computed. `SYSTEM_INDEX_DESCRIPTORS` is used here in the [SystemIndices](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndices.java#L94-L96) class where a list is a more suitable data structure. 

Ultimately, the map data structure introduced here is needed for the implementation of `matchesPluginSystemIndexPattern`.

I can modify everything to use the map, but there will be a lot of repeated computations.

Reply:
2 things here:
 - the SYSTEM_INDEX_DESCRIPTORS_MAP is missing `TASK_INDEX_DESCRIPTOR` (which was added to `descriptors` but is not present in the `descriptorsMap`)
 - flatMap() over SYSTEM_INDEX_DESCRIPTORS_MAP.values() should give us SYSTEM_INDEX_DESCRIPTORS_MAP so we need only  1 data structure 

How often do we need lookup by plugin class name?

Reply:
Hmm, you're right. [SystemIndices.validateSystemIndex](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndices.java#L128) is only called on create index so the call volume will not be huge. I think its reasonable to store the map and compute a list when needed.

Reply:
`descriptorsMap` does contain `TASK_INDEX_DESCRIPTOR`. 

1. `descriptorsMap` is built [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndexRegistry.java#L51) using `buildSystemIndexDescriptorMap`
2. [`buildSystemIndexDescriptorMap`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/SystemIndexRegistry.java#L113-L129) includes SERVER_SYSTEM_INDEX_DESCRIPTORS which includes the task index

Reply:
> How often do we need lookup by plugin class name?

Lookup would be required in blocks where a plugin has stashed the thread context and is performing an action against one of its system indices.

Reply:
> descriptorsMap does contain TASK_INDEX_DESCRIPTOR.

Indeed, you are right (the initialization is quite messy to be fair)

> I think its reasonable to store the map and compute a list when needed.

Let's do that?

Reply:
Done. Pushed a commit to consolidate to a single map data structure.